codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -213,33 +216,55 @@ class KotlinJmespathExpressionVisitor(
         return acceptSubexpression(this.arguments[0]) to acceptSubexpression(this.arguments[1])
     }
 
+    private fun VisitedExpression.dotFunction(expression: FunctionExpression, expr: String, elvisExpr: String? = null): VisitedExpression {
+        val dotFunctionExpr = ensureNullGuard(shape, expr, elvisExpr)
+        val ident = addTempVar(expression.name, "$identifier$dotFunctionExpr")
+
+        return VisitedExpression(ident, shape)
+    }
+
     override fun visitFunction(expression: FunctionExpression): VisitedExpression = when (expression.name) {
         "contains" -> {
             val (subject, search) = expression.twoArgs()
-
-            val containsExpr = ensureNullGuard(subject.shape, "contains(${search.identifier})", "false")
-            val ident = addTempVar("contains", "${subject.identifier}$containsExpr")
-
-            VisitedExpression(ident)
+            subject.dotFunction(expression, "contains(${search.identifier})", "false")
         }
 
         "length" -> {
             writer.addImport(RuntimeTypes.Core.Utils.length)
             val subject = expression.singleArg()
-
-            val lengthExpr = ensureNullGuard(subject.shape, "length", "0")
-            val ident = addTempVar("length", "${subject.identifier}$lengthExpr")
-
-            VisitedExpression(ident)
+            subject.dotFunction(expression, "length", "0")
         }
 
         "abs", "floor", "ceil" -> {
             val number = expression.singleArg()
+            number.dotFunction(expression, "let { kotlin.math.${expression.name}(it.toDouble()) }")
+        }
+
+        "sum" -> {
+            val numbers = expression.singleArg()
+            numbers.dotFunction(expression, "sum()")
+        }
+
+        "avg" -> {
+            val numbers = expression.singleArg()
+            val average = numbers.dotFunction(expression, "average()").identifier
+            val isNaN = ensureNullGuard(numbers.shape, "isNaN() == true")
+            VisitedExpression(addTempVar("averageOrNull", "if($average$isNaN) null else $average"), nullable = true)
+        }
```

Comment:
**Suggestion**: [`takeUnless`](https://kotlinlang.org/docs/scope-functions.html#takeif-and-takeunless) can simplify this for you:

```kotlin
"avg" -> {
    val numbers = expression.singleArg()
    numbers.dotFunction(expression, "average()").dotFunction(expression, "takeUnless { it.isNan() }")
}
```

Chaining `dotFunction` twice with the same `expression` is a little cheesy because it would result in codegen like:

```kotlin
val average = theList?.average()
val average2 = average?.takeUnless { it.isNan() }
```

If you go this route, it might be worth adding an overload of `dotFunction` that can take a string:

```kotlin
private fun VisitedExpression.dotFunction(preferredName: String, expr: String, elvisExpr: String? = null): VisitedExpression {
    val dotFunctionExpr = ensureNullGuard(shape, expr, elvisExpr)
    val ident = addTempVar(preferredName, "$identifier$dotFunctionExpr")
    
    return VisitedExpression(ident, shape)
}

private fun VisitedExpression.dotFunction(expression: FunctionExpression, expr: String, elvisExpr: String? = null) =
    dotFunction(expression.name, expr, elvisExpr)

...

"avg" -> {
    val numbers = expression.singleArg()
    numbers.dotFunction(expression, "average()").dotFunction("averageOrNull, "takeUnless { it.isNan() }")
}
```